micrometer-core/src/main/java/io/micrometer/core/instrument/binder/kafka/KafkaMetrics.java

```diff
@@ -0,0 +1,279 @@
+        this(kafkaProducer::metrics, tags);
+    }
+
+    /**
+     * Kafka {@link Producer} metrics binder
+     *
+     * @param kafkaProducer producer instance to be instrumented
+     */
+    public KafkaMetrics(Producer<?, ?> kafkaProducer) {
+        this(kafkaProducer::metrics);
+    }
+
+    /**
+     * Kafka {@link Consumer} metrics binder
+     *
+     * @param kafkaConsumer consumer instance to be instrumented
+     * @param tags          additional tags
+     */
+    public KafkaMetrics(Consumer<?, ?> kafkaConsumer, Iterable<Tag> tags) {
+        this(kafkaConsumer::metrics, tags);
+    }
+
+    /**
+     * Kafka {@link Consumer} metrics binder
+     *
+     * @param kafkaConsumer consumer instance to be instrumented
+     */
+    public KafkaMetrics(Consumer<?, ?> kafkaConsumer) {
+        this(kafkaConsumer::metrics);
+    }
+
+    /**
+     * {@link KafkaStreams} metrics binder
+     *
+     * @param kafkaStreams instance to be instrumented
+     * @param tags         additional tags
+     */
+    public KafkaMetrics(KafkaStreams kafkaStreams, Iterable<Tag> tags) {
+        this(kafkaStreams::metrics, tags);
+    }
+
+    /**
+     * {@link KafkaStreams} metrics binder
+     *
+     * @param kafkaStreams instance to be instrumented
+     */
+    public KafkaMetrics(KafkaStreams kafkaStreams) {
+        this(kafkaStreams::metrics);
+    }
+
+    /**
+     * Kafka {@link AdminClient} metrics binder
+     *
+     * @param adminClient instance to be instrumented
+     * @param tags        additional tags
+     */
+    public KafkaMetrics(AdminClient adminClient, Iterable<Tag> tags) {
+        this(adminClient::metrics, tags);
+    }
+
+    /**
+     * Kafka {@link AdminClient} metrics binder
+     *
+     * @param adminClient instance to be instrumented
+     */
+    public KafkaMetrics(AdminClient adminClient) {
+        this(adminClient::metrics);
+    }
+
+    KafkaMetrics(Supplier<Map<MetricName, ? extends Metric>> metricsSupplier) {
+        this(metricsSupplier, emptyList());
+    }
+
+    KafkaMetrics(Supplier<Map<MetricName, ? extends Metric>> metricsSupplier,
+            Iterable<Tag> extraTags) {
+        this.metricsSupplier = metricsSupplier;
+        this.extraTags = extraTags;
+    }
+
+    @Override
+    public void bindTo(MeterRegistry registry) {
+        checkAndBindMetrics(registry);
+    }
+
+    /**
+     * Gather metrics from Kafka metrics API and register Meters.
+     * <p>
+     * As this is a one-off execution when binding a Kafka client, Meters include a call to this
+     * validation to double-check new metrics when returning values. This should only add the cost of
+     * validating meters registered counter when no new meters are present.
+     */
+    void checkAndBindMetrics(MeterRegistry registry) {
+        Map<MetricName, ? extends Metric> metrics = metricsSupplier.get();
+        //Only happens first time number of metrics change
+        if (currentSize.get() != metrics.size()) {
```

Comment:
That's true. I was only considering metrics number is increasing over time, under the assumption that we have only one client (producer/consumer/streams) per `KafkaMetrics` instance, and clients will send/poll an increasing number of topics.

Another option would be to keep a set of `MetricName`s and compare both sets instead of just the number; maybe a bit more expensive but more correct as well.